pkg/endpoints/manager.go

```diff
@@ -28,19 +31,75 @@ type Manager struct {
 
 	EndpointSizeCallback func(deploymentName string, size int)
 
-	endpointsMtx sync.Mutex
-	endpoints    map[string]*endpointGroup
+	// service name to endpoints
+	svcEndpointsMtx sync.Mutex
+	svcEndpoints    map[string]*endpointGroup
+	// model to endpoints
+	podEndpointsMtx sync.Mutex
+	podEndpoints    map[string]*endpointGroup
 
 	ExcludePods map[string]struct{}
 }
 
 func (r *Manager) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
+	if err := ctrl.NewControllerManagedBy(mgr).
 		For(&disv1.EndpointSlice{}).
-		Complete(r)
+		Complete(ReconcilerFn(r.ReconcileSVC)); err != nil {
+		return ctrl.NewControllerManagedBy(mgr).
+			For(&disv1.EndpointSlice{}).
+			Complete(ReconcilerFn(r.ReconcileSVC))
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		Complete(ReconcilerFn(r.ReconcilePods))
+}
+
+func (r *Manager) ReconcilePods(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var pod corev1.Pod
+	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	switch pod.Status.Phase {
+	case corev1.PodPending:
+		return ctrl.Result{}, nil
+	case corev1.PodRunning:
+		// new or updated instance
+		ports := make(map[string]int32)
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				ports[port.Name] = port.ContainerPort
```

Comment:
Question to myself, in K8s can you call the port of a container without it being defined as a ContainerPort?

Reply:
I just tested this, seems like you can.  So purely relying on pods to get the port might not be correct either.

Here is the test I ran:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-nginx
spec:
  selector:
    matchLabels:
      run: my-nginx
  replicas: 2
  template:
    metadata:
      labels:
        run: my-nginx
    spec:
      containers:
      - name: my-nginx
        image: nginx
```

Create a svc:
```
  ports:
  - port: 80
    protocol: TCP
    targetPort: 80
  selector:
    run: my-nginx
```

Check the endpoints:
```
my-nginx     10.244.0.5:80,10.244.0.6:80   2m49s
```

Notice that endpoint does have the correct port but pod information does not:
```
  containers:                                           
  - image: nginx                                        
    imagePullPolicy: Always                             
    name: my-nginx      
    resources: {}                                                                                               
    terminationMessagePath: /dev/termination-log        
    terminationMessagePolicy: File                                                                              
    volumeMounts:                                       
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-dsn4b                       
      readOnly: true 
```

Reply:
Good catch. Not sure what is the best approach in this scenario. I think it would be fair to log an error and ignore the pod but I am curious for other options
Btw, with the pods as source for IPs, we don't need the service anymore.